pkg/kv/kvserver/replica_proposal_buf.go

```diff
@@ -456,61 +461,170 @@ func (b *propBuf) FlushLockedWithRaftGroup(
+			entryEncoding = raftlog.EntryEncodingSideloaded
+		case p.command.ReplicatedEvalResult.ChangeReplicas != nil:
+			// Flush any previously batched (non-conf change) proposals to
+			// preserve the correct ordering or proposals. This isn't
+			// strictly necessary for anything - after all, latching prevents
+			// reordering where it matters, but it's not worth changing now.
+			// Later proposals will start a new batch.
+			if err := proposeBatch(raftGroup, b.p.getReplicaID(), ents); err != nil {
+				firstErr = err
+				continue
+			}
+			ents = ents[len(ents):]
+
+			checkCC, err := p.command.ReplicatedEvalResult.ChangeReplicas.ConfChange(nil /* encodedCtx */)
+			if err != nil {
+				firstErr = err
+				continue
+			}
+			if _, v1 := checkCC.AsV1(); v1 {
+				entryEncoding = raftlog.EntryEncodingRaftConfChange
+			} else {
+				entryEncoding = raftlog.EntryEncodingRaftConfChangeV2
+			}
+		default:
+			entryEncoding = raftlog.EntryEncodingStandard
+		}
+
+		var ct hlc.Timestamp
 		// Figure out what closed timestamp this command will carry.
 		//
 		// If this is a reproposal, we don't reassign the LAI. We also don't
 		// reassign the closed timestamp: we could, in principle, but we'd have to
 		// make a copy of the encoded command as to not modify the copy that's
 		// already stored in the local replica's raft entry cache.
 		if !reproposal {
-			lai, closedTimestamp, err := b.allocateLAIAndClosedTimestampLocked(ctx, p, closedTSTarget)
-			if err != nil {
-				firstErr = err
-				continue
-			}
-			err = b.marshallLAIAndClosedTimestampToProposalLocked(ctx, p, lai, closedTimestamp)
+			var lai uint64
+			var err error
+			lai, ct, err = b.allocateLAIAndClosedTimestampLocked(ctx, p, closedTSTarget)
 			if err != nil {
 				firstErr = err
 				continue
 			}
+			// Also assign MaxLeaseIndex to the in-memory copy. The in-memory copy is
+			// checked for sanity at application time, on the proposing replica. Note
+			// that if we catch any error below, this proposal will be rejected, so
+			// we can do this upfront.
+			p.command.MaxLeaseIndex = lai
+			// NB: we don't assign to the p.command.ClosedTimestamp here because that
+			// would cause an allocation. It'd be nice to assign to it, for
+			// consistency, but nobody needs it. What we do instead is get memory from
+			// a sync.Pool just while we're marshaling the command, and reset the
+			// field again right after. We could make the field nullable, see comment
+			// on p.command.ClosedTimestamp for context.
+			_ = ct
 		}
 
-		// Potentially drop the proposal before passing it to etcd/raft, but
-		// only after performing necessary bookkeeping.
+		testingSubmitProposalFilter := func(*ProposalData) (drop bool, err error) { return false, nil }
 		if filter := b.testing.submitProposalFilter; filter != nil {
-			if drop, err := filter(p); drop || err != nil {
+			testingSubmitProposalFilter = filter
+		}
+
+		// TODO(tbg): the encoding code within both of the branches of this
```

Comment:
Probably a good idea to do it in this PR (not necessarily in `logstore`, it's fine to have a helper in this file, and a TODO to refine and move it elsewhere).

reflection: Not inlining this bit would simplify reviewing the diff in `Flush`.